# SaaS Churn & Revenue Analysis — Data Cleaning & EDA

## Purpose

This notebook performs a beginner-friendly data understanding and data quality review for the RavenStack synthetic SaaS dataset.

Main goals:

1. Load all raw CSV files.
2. Check table shapes and row counts.
3. Review column names and data types.
4. Identify missing values and duplicate keys.
5. Prepare cleaned data for SQL and Power BI.

In [85]:
import pandas as pd

pd.set_option("display.max_columns", 50)
pd.set_option("display.width", 120)

print("Pandas version:", pd.__version__)

Pandas version: 2.2.2


In [86]:
base_url = "https://raw.githubusercontent.com/PavelBoyko-beep/data-analytics-portfolio/main/saas-churn-revenue-analysis/data/raw"

accounts = pd.read_csv(f"{base_url}/ravenstack_accounts.csv")
subscriptions = pd.read_csv(f"{base_url}/ravenstack_subscriptions.csv")
feature_usage = pd.read_csv(f"{base_url}/ravenstack_feature_usage.csv")
support_tickets = pd.read_csv(f"{base_url}/ravenstack_support_tickets.csv")
churn_events = pd.read_csv(f"{base_url}/ravenstack_churn_events.csv")

print("All CSV files loaded successfully.")

All CSV files loaded successfully.


In [87]:
tables = {
    "accounts": accounts,
    "subscriptions": subscriptions,
    "feature_usage": feature_usage,
    "support_tickets": support_tickets,
    "churn_events": churn_events
}

row_counts = pd.DataFrame({
    "table_name": list(tables.keys()),
    "row_count": [len(df) for df in tables.values()]
})

row_counts

,table_name,row_count
0,accounts,500
1,subscriptions,5000
2,feature_usage,25000
3,support_tickets,2000
4,churn_events,600


In [88]:
tables = {
    "accounts": accounts,
    "subscriptions": subscriptions,
    "feature_usage": feature_usage,
    "support_tickets": support_tickets,
    "churn_events": churn_events
}

for table_name, df in tables.items():
    print("=" * 60)
    print(f"TABLE: {table_name}")
    print("=" * 60)

    print("\nShape:")
    print(df.shape)

    print("\nFirst 5 rows:")
    display(df.head())

    print("\nColumn names:")
    print(df.columns.tolist())

    print("\n")

TABLE: accounts

Shape:
(500, 10)

First 5 rows:


,account_id,account_name,industry,country,signup_date,referral_source,plan_tier,seats,is_trial,churn_flag
0,A-2e4581,Company_0,EdTech,US,2024-10-16,partner,Basic,9,False,False
1,A-43a9e3,Company_1,FinTech,IN,2023-08-17,other,Basic,18,False,True
2,A-0a282f,Company_2,DevTools,US,2024-08-27,organic,Basic,1,False,False
3,A-1f0ac7,Company_3,HealthTech,UK,2023-08-27,other,Basic,24,True,False
4,A-ce550d,Company_4,HealthTech,US,2024-10-27,event,Enterprise,35,False,True



Column names:
['account_id', 'account_name', 'industry', 'country', 'signup_date', 'referral_source', 'plan_tier', 'seats', 'is_trial', 'churn_flag']


TABLE: subscriptions

Shape:
(5000, 14)

First 5 rows:


,subscription_id,account_id,start_date,end_date,plan_tier,seats,mrr_amount,arr_amount,is_trial,upgrade_flag,downgrade_flag,churn_flag,billing_frequency,auto_renew_flag
0,S-8cec59,A-3c1a3f,2023-12-23,2024-04-12,Enterprise,14,2786,33432,False,False,False,True,monthly,True
1,S-0f6f44,A-9b9fe9,2024-06-11,NaN,Pro,17,833,9996,False,False,False,False,monthly,True
2,S-51c0d1,A-659280,2024-11-25,NaN,Enterprise,62,0,0,True,True,False,False,annual,False
3,S-f81687,A-e7a1e2,2024-11-23,2024-12-13,Enterprise,5,995,11940,False,False,False,True,monthly,True
4,S-cff5a2,A-ba6516,2024-01-10,NaN,Enterprise,27,5373,64476,False,False,False,False,monthly,True



Column names:
['subscription_id', 'account_id', 'start_date', 'end_date', 'plan_tier', 'seats', 'mrr_amount', 'arr_amount', 'is_trial', 'upgrade_flag', 'downgrade_flag', 'churn_flag', 'billing_frequency', 'auto_renew_flag']


TABLE: feature_usage

Shape:
(25000, 8)

First 5 rows:


,usage_id,subscription_id,usage_date,feature_name,usage_count,usage_duration_secs,error_count,is_beta_feature
0,U-1c6c24,S-0fcf7d,2023-07-27,feature_20,9,5004,0,False
1,U-f07cb8,S-c25263,2023-08-07,feature_5,9,369,0,False
2,U-096807,S-f29e7f,2023-12-07,feature_3,9,1458,0,False
3,U-6b1580,S-be655e,2024-07-28,feature_40,5,2085,0,False
4,U-720a29,S-f9b1d0,2024-12-02,feature_12,12,900,0,False



Column names:
['usage_id', 'subscription_id', 'usage_date', 'feature_name', 'usage_count', 'usage_duration_secs', 'error_count', 'is_beta_feature']


TABLE: support_tickets

Shape:
(2000, 9)

First 5 rows:


,ticket_id,account_id,submitted_at,closed_at,resolution_time_hours,priority,first_response_time_minutes,satisfaction_score,escalation_flag
0,T-0024de,A-712f1c,2023-07-27,2023-07-28 03:00:00,27.0,high,74,NaN,False
1,T-4d04b9,A-e43bf7,2024-07-08,2024-07-09 03:00:00,27.0,urgent,144,NaN,False
2,T-d5e12f,A-0f3e88,2024-10-17,2024-10-17 19:00:00,19.0,urgent,93,4.0,False
3,T-dfce9a,A-4c56c9,2024-09-08,2024-09-09 23:00:00,47.0,medium,126,5.0,False
4,T-c59f77,A-6f8ad2,2024-11-30,2024-12-01 02:00:00,26.0,medium,8,NaN,False



Column names:
['ticket_id', 'account_id', 'submitted_at', 'closed_at', 'resolution_time_hours', 'priority', 'first_response_time_minutes', 'satisfaction_score', 'escalation_flag']


TABLE: churn_events

Shape:
(600, 9)

First 5 rows:


,churn_event_id,account_id,churn_date,reason_code,refund_amount_usd,preceding_upgrade_flag,preceding_downgrade_flag,is_reactivation,feedback_text
0,C-816288,A-c37cab,2024-10-27,pricing,4.03,False,False,False,switched to competitor
1,C-5a81e7,A-37f969,2024-06-25,support,96.45,True,False,False,NaN
2,C-a174be,A-b07346,2024-11-12,budget,0.00,False,False,False,missing features
3,C-accb39,A-1e50e0,2023-11-01,budget,54.94,False,False,False,switched to competitor
4,C-92f889,A-956988,2024-12-30,unknown,0.00,False,True,True,too expensive



Column names:
['churn_event_id', 'account_id', 'churn_date', 'reason_code', 'refund_amount_usd', 'preceding_upgrade_flag', 'preceding_downgrade_flag', 'is_reactivation', 'feedback_text']




In [89]:
for table_name, df in tables.items():

    print("=" * 60)
    print(table_name.upper())
    print("=" * 60)

    df.info()

    print("\n")

ACCOUNTS
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   account_id       500 non-null    object
 1   account_name     500 non-null    object
 2   industry         500 non-null    object
 3   country          500 non-null    object
 4   signup_date      500 non-null    object
 5   referral_source  500 non-null    object
 6   plan_tier        500 non-null    object
 7   seats            500 non-null    int64 
 8   is_trial         500 non-null    bool  
 9   churn_flag       500 non-null    bool  
dtypes: bool(2), int64(1), object(7)
memory usage: 32.4+ KB


SUBSCRIPTIONS
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   subscription_id    5000 non-null   object
 1   account_id       

In [90]:
missing_values = pd.DataFrame()

for table_name, df in tables.items():

    missing_count = df.isnull().sum()

    missing_percent = (
        df.isnull().sum() / len(df) * 100
    ).round(2)

    temp_df = pd.DataFrame({
        "column_name": missing_count.index,
        "missing_count": missing_count.values,
        "missing_percent": missing_percent.values
    })

    temp_df["table_name"] = table_name

    missing_values = pd.concat(
        [missing_values, temp_df],
        ignore_index=True
    )

missing_values = (
    missing_values
    .sort_values(
        by="missing_count",
        ascending=False
    )
)

missing_values.head(20)

,column_name,missing_count,missing_percent,table_name
13,end_date,4514,90.28,subscriptions
39,satisfaction_score,825,41.25,support_tickets
49,feedback_text,148,24.67,churn_events
2,industry,0,0.00,accounts
1,account_name,0,0.00,accounts
3,country,0,0.00,accounts
6,plan_tier,0,0.00,accounts
7,seats,0,0.00,accounts
4,signup_date,0,0.00,accounts
5,referral_source,0,0.00,accounts


In [91]:
duplicate_rows = pd.DataFrame()

for table_name, df in tables.items():

    duplicate_count = df.duplicated().sum()

    temp_df = pd.DataFrame({
        "table_name": [table_name],
        "duplicate_rows": [duplicate_count],
        "total_rows": [len(df)]
    })

    temp_df["duplicate_percent"] = (
        temp_df["duplicate_rows"] / temp_df["total_rows"] * 100
    ).round(2)

    duplicate_rows = pd.concat(
        [duplicate_rows, temp_df],
        ignore_index=True
    )

duplicate_rows

,table_name,duplicate_rows,total_rows,duplicate_percent
0,accounts,0,500,0.0
1,subscriptions,0,5000,0.0
2,feature_usage,0,25000,0.0
3,support_tickets,0,2000,0.0
4,churn_events,0,600,0.0


In [92]:
primary_keys = {
    "accounts": "account_id",
    "subscriptions": "subscription_id",
    "feature_usage": "usage_id",
    "support_tickets": "ticket_id",
    "churn_events": "churn_event_id"
}

primary_key_checks = pd.DataFrame()

for table_name, key_column in primary_keys.items():

    df = tables[table_name]

    total_rows = len(df)
    unique_keys = df[key_column].nunique()
    duplicate_keys = total_rows - unique_keys
    missing_keys = df[key_column].isnull().sum()

    temp_df = pd.DataFrame({
        "table_name": [table_name],
        "primary_key": [key_column],
        "total_rows": [total_rows],
        "unique_keys": [unique_keys],
        "duplicate_keys": [duplicate_keys],
        "missing_keys": [missing_keys]
    })

    primary_key_checks = pd.concat(
        [primary_key_checks, temp_df],
        ignore_index=True
    )

primary_key_checks

,table_name,primary_key,total_rows,unique_keys,duplicate_keys,missing_keys
0,accounts,account_id,500,500,0,0
1,subscriptions,subscription_id,5000,5000,0,0
2,feature_usage,usage_id,25000,24979,21,0
3,support_tickets,ticket_id,2000,2000,0,0
4,churn_events,churn_event_id,600,600,0,0


In [93]:
duplicate_usage_ids = (
    feature_usage[feature_usage["usage_id"].duplicated(keep=False)]
    .sort_values("usage_id")
)

duplicate_usage_ids

,usage_id,subscription_id,usage_date,feature_name,usage_count,usage_duration_secs,error_count,is_beta_feature
19294,U-0c9318,S-01b2dc,2023-12-11,feature_9,5,1060,3,False
17533,U-0c9318,S-0ffab0,2024-01-30,feature_11,3,1614,0,False
20588,U-13ce5b,S-9b623b,2023-03-26,feature_28,10,2050,1,False
9626,U-13ce5b,S-8b0950,2024-09-10,feature_9,8,824,0,True
18480,U-2103bb,S-7fc49b,2023-04-18,feature_12,9,5022,1,False
10379,U-2103bb,S-ae3270,2024-01-06,feature_3,10,5510,0,False
22,U-25b56c,S-810c27,2024-10-06,feature_20,7,231,0,False
7574,U-25b56c,S-34253c,2023-10-28,feature_20,6,2166,1,False
21376,U-48a4aa,S-93f835,2024-08-24,feature_39,10,4770,0,False
1085,U-48a4aa,S-383ac2,2023-02-12,feature_28,14,7126,0,False


In [94]:
duplicate_usage_summary = (
    duplicate_usage_ids
    .groupby("usage_id")
    .size()
    .reset_index(name="row_count")
    .sort_values("row_count", ascending=False)
)

duplicate_usage_summary

,usage_id,row_count
0,U-0c9318,2
1,U-13ce5b,2
2,U-2103bb,2
3,U-25b56c,2
4,U-48a4aa,2
5,U-4ef9da,2
6,U-52f4ae,2
7,U-5cd3c2,2
8,U-649cec,2
9,U-662254,2


## Primary Key Check Notes

The primary key check found that `usage_id` in the `feature_usage` table is not fully unique.

- Total rows in `feature_usage`: 25,000
- Unique `usage_id` values: 24,979
- Duplicate `usage_id` values: 21
- Missing `usage_id` values: 0

The duplicated `usage_id` values belong to different usage events, so these rows should not be removed.

Decision:
For SQL import, `usage_id` will be kept as the original raw identifier, but it will not be used as the primary key. A technical primary key should be created for the `feature_usage` table during PostgreSQL setup.

In [95]:
foreign_key_checks = []

# 1. subscriptions.account_id should exist in accounts.account_id
missing_subscription_accounts = subscriptions[
    ~subscriptions["account_id"].isin(accounts["account_id"])
]

foreign_key_checks.append({
    "relationship": "subscriptions.account_id -> accounts.account_id",
    "missing_foreign_keys": len(missing_subscription_accounts)
})

# 2. support_tickets.account_id should exist in accounts.account_id
missing_ticket_accounts = support_tickets[
    ~support_tickets["account_id"].isin(accounts["account_id"])
]

foreign_key_checks.append({
    "relationship": "support_tickets.account_id -> accounts.account_id",
    "missing_foreign_keys": len(missing_ticket_accounts)
})

# 3. churn_events.account_id should exist in accounts.account_id
missing_churn_accounts = churn_events[
    ~churn_events["account_id"].isin(accounts["account_id"])
]

foreign_key_checks.append({
    "relationship": "churn_events.account_id -> accounts.account_id",
    "missing_foreign_keys": len(missing_churn_accounts)
})

# 4. feature_usage.subscription_id should exist in subscriptions.subscription_id
missing_usage_subscriptions = feature_usage[
    ~feature_usage["subscription_id"].isin(subscriptions["subscription_id"])
]

foreign_key_checks.append({
    "relationship": "feature_usage.subscription_id -> subscriptions.subscription_id",
    "missing_foreign_keys": len(missing_usage_subscriptions)
})

foreign_key_checks = pd.DataFrame(foreign_key_checks)

foreign_key_checks

,relationship,missing_foreign_keys
0,subscriptions.account_id -> accounts.account_id,0
1,support_tickets.account_id -> accounts.account_id,0
2,churn_events.account_id -> accounts.account_id,0
3,feature_usage.subscription_id -> subscriptions...,0


In [96]:
date_columns = {
    "accounts": ["signup_date"],
    "subscriptions": ["start_date", "end_date"],
    "feature_usage": ["usage_date"],
    "support_tickets": ["submitted_at", "closed_at"],
    "churn_events": ["churn_date"]
}

for table_name, columns in date_columns.items():
    df = tables[table_name]

    for column in columns:
        df[column] = pd.to_datetime(df[column], errors="coerce")

print("Date conversion completed.")

Date conversion completed.


In [97]:
for table_name, columns in date_columns.items():
    print("=" * 60)
    print(table_name.upper())
    print("=" * 60)

    for column in columns:
        print(column, "->", tables[table_name][column].dtype)

    print("\n")

ACCOUNTS
signup_date -> datetime64[ns]


SUBSCRIPTIONS
start_date -> datetime64[ns]
end_date -> datetime64[ns]


FEATURE_USAGE
usage_date -> datetime64[ns]


SUPPORT_TICKETS
submitted_at -> datetime64[ns]
closed_at -> datetime64[ns]


CHURN_EVENTS
churn_date -> datetime64[ns]




In [98]:
date_logic_checks = []

# 1. Subscription end date should not be before start date
invalid_subscription_dates = subscriptions[
    subscriptions["end_date"].notna() &
    (subscriptions["end_date"] < subscriptions["start_date"])
]

date_logic_checks.append({
    "check_name": "subscription end_date before start_date",
    "invalid_rows": len(invalid_subscription_dates)
})

# 2. Subscription start date should not be before account signup date
subscriptions_with_accounts = subscriptions.merge(
    accounts[["account_id", "signup_date"]],
    on="account_id",
    how="left"
)

invalid_subscription_before_signup = subscriptions_with_accounts[
    subscriptions_with_accounts["start_date"] < subscriptions_with_accounts["signup_date"]
]

date_logic_checks.append({
    "check_name": "subscription start_date before account signup_date",
    "invalid_rows": len(invalid_subscription_before_signup)
})

# 3. Ticket closed_at should not be before submitted_at
invalid_ticket_dates = support_tickets[
    support_tickets["closed_at"] < support_tickets["submitted_at"]
]

date_logic_checks.append({
    "check_name": "ticket closed_at before submitted_at",
    "invalid_rows": len(invalid_ticket_dates)
})

# 4. Churn date should not be before account signup date
churn_with_accounts = churn_events.merge(
    accounts[["account_id", "signup_date"]],
    on="account_id",
    how="left"
)

invalid_churn_before_signup = churn_with_accounts[
    churn_with_accounts["churn_date"] < churn_with_accounts["signup_date"]
]

date_logic_checks.append({
    "check_name": "churn_date before account signup_date",
    "invalid_rows": len(invalid_churn_before_signup)
})

# 5. Feature usage should not happen before subscription start date
usage_with_subscriptions = feature_usage.merge(
    subscriptions[["subscription_id", "start_date"]],
    on="subscription_id",
    how="left"
)

invalid_usage_before_subscription = usage_with_subscriptions[
    usage_with_subscriptions["usage_date"] < usage_with_subscriptions["start_date"]
]

date_logic_checks.append({
    "check_name": "usage_date before subscription start_date",
    "invalid_rows": len(invalid_usage_before_subscription)
})

date_logic_checks = pd.DataFrame(date_logic_checks)

date_logic_checks

,check_name,invalid_rows
0,subscription end_date before start_date,0
1,subscription start_date before account signup_...,0
2,ticket closed_at before submitted_at,0
3,churn_date before account signup_date,0
4,usage_date before subscription start_date,19142


In [99]:
usage_date_diagnostics = usage_with_subscriptions.copy()

usage_date_diagnostics["days_from_subscription_start"] = (
    usage_date_diagnostics["usage_date"] - usage_date_diagnostics["start_date"]
).dt.days

usage_date_diagnostics[[
    "usage_id",
    "subscription_id",
    "usage_date",
    "start_date",
    "days_from_subscription_start",
    "feature_name",
    "usage_count",
    "usage_duration_secs",
    "error_count",
    "is_beta_feature"
]].sort_values("days_from_subscription_start").head(20)

,usage_id,subscription_id,usage_date,start_date,days_from_subscription_start,feature_name,usage_count,usage_duration_secs,error_count,is_beta_feature
21121,U-687d44,S-ad31b0,2023-01-01,2024-12-29,-728,feature_30,7,959,1,False
13779,U-34991b,S-b57a5e,2023-01-02,2024-12-30,-728,feature_20,9,3384,0,False
5633,U-d98ada,S-6c6e62,2023-01-02,2024-12-29,-727,feature_20,13,169,1,False
2876,U-e4fa89,S-2f11bb,2023-01-02,2024-12-29,-727,feature_33,7,3486,0,False
4621,U-860f23,S-6fa3a3,2023-01-04,2024-12-30,-726,feature_38,10,740,0,False
2198,U-7bf6f0,S-9c414f,2023-01-01,2024-12-27,-726,feature_30,10,3030,1,False
23405,U-8ca16d,S-7077ab,2023-01-06,2024-12-30,-724,feature_22,9,1377,0,False
14072,U-a9eaa6,S-c15f06,2023-01-04,2024-12-27,-723,feature_7,11,1430,0,False
13636,U-032738,S-240a09,2023-01-07,2024-12-30,-723,feature_21,10,4940,3,False
16088,U-a3fd7a,S-58a7c0,2023-01-01,2024-12-24,-723,feature_7,6,2592,1,True


In [100]:
usage_date_diagnostics["days_from_subscription_start"].describe()

,days_from_subscription_start
count,25000.000000
mean,-196.074800
std,261.387345
min,-728.000000
25%,-398.000000
50%,-197.000000
75%,-13.000000
max,692.000000


In [101]:
usage_timing_summary = pd.DataFrame({
    "category": [
        "usage before subscription start",
        "usage on/after subscription start"
    ],
    "row_count": [
        (usage_date_diagnostics["days_from_subscription_start"] < 0).sum(),
        (usage_date_diagnostics["days_from_subscription_start"] >= 0).sum()
    ]
})

usage_timing_summary["percent"] = (
    usage_timing_summary["row_count"] / len(usage_date_diagnostics) * 100
).round(2)

usage_timing_summary

,category,row_count,percent
0,usage before subscription start,19142,76.57
1,usage on/after subscription start,5858,23.43


In [102]:
first_usage_by_subscription = (
    feature_usage
    .groupby("subscription_id", as_index=False)
    .agg(
        first_usage_date=("usage_date", "min"),
        last_usage_date=("usage_date", "max"),
        usage_events=("usage_id", "count")
    )
)

subscription_usage_timing = subscriptions.merge(
    first_usage_by_subscription,
    on="subscription_id",
    how="left"
)

subscription_usage_timing["days_first_usage_to_subscription_start"] = (
    subscription_usage_timing["first_usage_date"] - subscription_usage_timing["start_date"]
).dt.days

subscription_usage_timing[[
    "subscription_id",
    "account_id",
    "start_date",
    "end_date",
    "first_usage_date",
    "last_usage_date",
    "days_first_usage_to_subscription_start",
    "usage_events",
    "is_trial",
    "billing_frequency",
    "plan_tier"
]].head(20)

,subscription_id,account_id,start_date,end_date,first_usage_date,last_usage_date,days_first_usage_to_subscription_start,usage_events,is_trial,billing_frequency,plan_tier
0,S-8cec59,A-3c1a3f,2023-12-23,2024-04-12,2023-07-14,2024-10-06,-162.0,6.0,False,monthly,Enterprise
1,S-0f6f44,A-9b9fe9,2024-06-11,NaT,2024-06-08,2024-11-30,-3.0,4.0,False,monthly,Pro
2,S-51c0d1,A-659280,2024-11-25,NaT,2023-07-29,2024-12-18,-485.0,3.0,True,annual,Enterprise
3,S-f81687,A-e7a1e2,2024-11-23,2024-12-13,2023-01-25,2024-12-10,-668.0,7.0,False,monthly,Enterprise
4,S-cff5a2,A-ba6516,2024-01-10,NaT,NaT,NaT,NaN,NaN,False,monthly,Enterprise
5,S-4b9b13,A-fa2041,2024-08-13,NaT,2023-04-26,2024-11-15,-475.0,5.0,False,monthly,Pro
6,S-dceac6,A-417d2f,2023-12-30,NaT,2023-06-07,2024-09-18,-206.0,3.0,False,annual,Enterprise
7,S-8cad7b,A-5f2961,2024-12-23,NaT,2023-08-25,2024-09-24,-486.0,3.0,False,annual,Basic
8,S-24796e,A-cc8c8f,2024-11-27,NaT,2023-05-10,2024-06-14,-567.0,6.0,False,annual,Enterprise
9,S-d0c344,A-80eeb6,2024-10-27,NaT,2023-02-18,2024-12-23,-617.0,8.0,False,annual,Pro


In [103]:
first_usage_timing_summary = pd.DataFrame({
    "category": [
        "first usage before subscription start",
        "first usage on subscription start",
        "first usage after subscription start",
        "no usage events"
    ],
    "subscription_count": [
        (subscription_usage_timing["days_first_usage_to_subscription_start"] < 0).sum(),
        (subscription_usage_timing["days_first_usage_to_subscription_start"] == 0).sum(),
        (subscription_usage_timing["days_first_usage_to_subscription_start"] > 0).sum(),
        (subscription_usage_timing["first_usage_date"].isna()).sum()
    ]
})

first_usage_timing_summary["percent"] = (
    first_usage_timing_summary["subscription_count"] / len(subscription_usage_timing) * 100
).round(2)

first_usage_timing_summary

,category,subscription_count,percent
0,first usage before subscription start,4783,95.66
1,first usage on subscription start,0,0.00
2,first usage after subscription start,184,3.68
3,no usage events,33,0.66


In [104]:
subscription_usage_timing["days_first_usage_to_subscription_start"].describe()

,days_first_usage_to_subscription_start
count,4967.000000
mean,-421.031206
std,203.406651
min,-728.000000
25%,-583.000000
50%,-463.000000
75%,-298.000000
max,570.000000


In [105]:
subscription_usage_with_accounts = subscription_usage_timing.merge(
    accounts[["account_id", "signup_date"]],
    on="account_id",
    how="left"
)

subscription_usage_with_accounts["days_first_usage_to_account_signup"] = (
    subscription_usage_with_accounts["first_usage_date"] - subscription_usage_with_accounts["signup_date"]
).dt.days

first_usage_vs_signup_summary = pd.DataFrame({
    "category": [
        "first usage before account signup",
        "first usage on/after account signup",
        "no usage events"
    ],
    "subscription_count": [
        (subscription_usage_with_accounts["days_first_usage_to_account_signup"] < 0).sum(),
        (subscription_usage_with_accounts["days_first_usage_to_account_signup"] >= 0).sum(),
        (subscription_usage_with_accounts["first_usage_date"].isna()).sum()
    ]
})

first_usage_vs_signup_summary["percent"] = (
    first_usage_vs_signup_summary["subscription_count"] / len(subscription_usage_with_accounts) * 100
).round(2)

first_usage_vs_signup_summary

,category,subscription_count,percent
0,first usage before account signup,4122,82.44
1,first usage on/after account signup,845,16.90
2,no usage events,33,0.66


In [106]:
subscription_usage_with_accounts["days_first_usage_to_account_signup"].describe()

,days_first_usage_to_account_signup
count,4967.000000
mean,-248.000403
std,248.899106
min,-724.000000
25%,-446.500000
50%,-260.000000
75%,-63.500000
max,715.000000


## Feature Usage Date Limitation

The date logic checks found a major limitation in the `feature_usage` table.

Initial check:
- 19,142 usage events happened before the related subscription start date.
- This represents 76.57% of all feature usage records.

Additional validation:
- For 4,122 subscriptions, the first usage event happened before the related account signup date.
- This represents 82.44% of subscription records.

Interpretation:
This means `feature_usage.usage_date` should not be used for lifecycle sequencing, such as:
- time from signup to first usage;
- usage before/after subscription start;
- time-to-value analysis;
- cohort analysis based on first usage date.

Decision:
The `feature_usage` table will still be used for product activity analysis, such as feature usage volume, error counts, and beta feature usage. However, usage dates will not be used for business conclusions that depend on the order of signup, subscription start, and usage events.

In [107]:
numeric_checks = []

# accounts
numeric_checks.append({
    "table_name": "accounts",
    "column_name": "seats",
    "negative_values": (accounts["seats"] < 0).sum(),
    "zero_values": (accounts["seats"] == 0).sum(),
    "min_value": accounts["seats"].min(),
    "max_value": accounts["seats"].max()
})

# subscriptions
for column in ["seats", "mrr_amount", "arr_amount"]:
    numeric_checks.append({
        "table_name": "subscriptions",
        "column_name": column,
        "negative_values": (subscriptions[column] < 0).sum(),
        "zero_values": (subscriptions[column] == 0).sum(),
        "min_value": subscriptions[column].min(),
        "max_value": subscriptions[column].max()
    })

# feature_usage
for column in ["usage_count", "usage_duration_secs", "error_count"]:
    numeric_checks.append({
        "table_name": "feature_usage",
        "column_name": column,
        "negative_values": (feature_usage[column] < 0).sum(),
        "zero_values": (feature_usage[column] == 0).sum(),
        "min_value": feature_usage[column].min(),
        "max_value": feature_usage[column].max()
    })

# support_tickets
for column in ["resolution_time_hours", "first_response_time_minutes"]:
    numeric_checks.append({
        "table_name": "support_tickets",
        "column_name": column,
        "negative_values": (support_tickets[column] < 0).sum(),
        "zero_values": (support_tickets[column] == 0).sum(),
        "min_value": support_tickets[column].min(),
        "max_value": support_tickets[column].max()
    })

# churn_events
numeric_checks.append({
    "table_name": "churn_events",
    "column_name": "refund_amount_usd",
    "negative_values": (churn_events["refund_amount_usd"] < 0).sum(),
    "zero_values": (churn_events["refund_amount_usd"] == 0).sum(),
    "min_value": churn_events["refund_amount_usd"].min(),
    "max_value": churn_events["refund_amount_usd"].max()
})

numeric_checks = pd.DataFrame(numeric_checks)

numeric_checks

,table_name,column_name,negative_values,zero_values,min_value,max_value
0,accounts,seats,0,0,1.0,163.00
1,subscriptions,seats,0,0,1.0,189.00
2,subscriptions,mrr_amount,0,778,0.0,33830.00
3,subscriptions,arr_amount,0,778,0.0,405960.00
4,feature_usage,usage_count,0,2,0.0,26.00
5,feature_usage,usage_duration_secs,0,2,0.0,12696.00
6,feature_usage,error_count,0,17269,0.0,8.00
7,support_tickets,resolution_time_hours,0,0,1.0,72.00
8,support_tickets,first_response_time_minutes,0,0,1.0,180.00
9,churn_events,refund_amount_usd,0,458,0.0,392.92


In [108]:
zero_revenue_subscriptions = subscriptions[
    (subscriptions["mrr_amount"] == 0) |
    (subscriptions["arr_amount"] == 0)
]

zero_revenue_summary = (
    zero_revenue_subscriptions
    .groupby(["is_trial", "churn_flag", "plan_tier", "billing_frequency"])
    .size()
    .reset_index(name="subscription_count")
    .sort_values("subscription_count", ascending=False)
)

zero_revenue_summary

,is_trial,churn_flag,plan_tier,billing_frequency,subscription_count
3,True,False,Enterprise,monthly,132
4,True,False,Pro,annual,116
5,True,False,Pro,monthly,115
2,True,False,Enterprise,annual,115
0,True,False,Basic,annual,112
1,True,False,Basic,monthly,110
9,True,True,Enterprise,monthly,18
11,True,True,Pro,monthly,16
7,True,True,Basic,monthly,13
10,True,True,Pro,annual,12


In [109]:
zero_revenue_subscriptions[[
    "subscription_id",
    "account_id",
    "start_date",
    "end_date",
    "plan_tier",
    "seats",
    "mrr_amount",
    "arr_amount",
    "is_trial",
    "churn_flag",
    "billing_frequency"
]].head(20)

,subscription_id,account_id,start_date,end_date,plan_tier,seats,mrr_amount,arr_amount,is_trial,churn_flag,billing_frequency
2,S-51c0d1,A-659280,2024-11-25,NaT,Enterprise,62,0,0,True,False,annual
11,S-79a9e1,A-3a5ad8,2024-12-25,NaT,Enterprise,22,0,0,True,False,annual
14,S-428e9a,A-8ae3fc,2024-10-12,NaT,Enterprise,88,0,0,True,False,monthly
16,S-13939b,A-86902e,2023-07-21,2024-08-03,Pro,34,0,0,True,True,monthly
20,S-c27134,A-dec005,2024-03-31,NaT,Pro,21,0,0,True,False,monthly
32,S-2877e6,A-95b24a,2024-06-28,NaT,Enterprise,38,0,0,True,False,monthly
34,S-d7d398,A-fb186e,2024-12-28,NaT,Enterprise,29,0,0,True,False,annual
35,S-ddad91,A-40906c,2024-06-21,NaT,Enterprise,92,0,0,True,False,annual
37,S-bb1cf1,A-3a5ad8,2024-09-07,NaT,Enterprise,9,0,0,True,False,monthly
49,S-ad39ef,A-e40f45,2024-08-25,NaT,Pro,21,0,0,True,False,annual


In [110]:
satisfaction_score_check = pd.DataFrame({
    "check_name": ["satisfaction_score outside 1-5"],
    "invalid_rows": [
        support_tickets[
            support_tickets["satisfaction_score"].notna() &
            ~support_tickets["satisfaction_score"].between(1, 5)
        ].shape[0]
    ],
    "missing_values": [support_tickets["satisfaction_score"].isna().sum()],
    "min_value": [support_tickets["satisfaction_score"].min()],
    "max_value": [support_tickets["satisfaction_score"].max()]
})

satisfaction_score_check

,check_name,invalid_rows,missing_values,min_value,max_value
0,satisfaction_score outside 1-5,0,825,3.0,5.0


## Data Quality Summary

The dataset was reviewed using Python/pandas before SQL import.

### Checks Completed

| Check | Result |
|---|---|
| Row counts | Passed — all 5 CSV files loaded with expected row counts |
| Missing values | Passed with notes |
| Duplicate rows | Passed — no fully duplicated rows found |
| Primary keys | Passed with note |
| Foreign keys | Passed — no orphan records found |
| Date type conversion | Passed |
| Date logic checks | Passed with limitation |
| Numeric value checks | Passed with notes |
| Satisfaction score validation | Passed |

### Key Findings

1. `subscriptions.end_date` has many missing values, but this is expected because active subscriptions do not have an end date.

2. `support_tickets.satisfaction_score` has missing values, but this is expected because not every customer leaves a satisfaction rating.

3. `churn_events.feedback_text` has missing values, but this is expected because not every churned customer provides written feedback.

4. The `feature_usage` table has 21 duplicated `usage_id` values. These rows are not full duplicates and should not be removed. For SQL import, a technical primary key should be created for the `feature_usage` table.

5. Foreign key checks passed. All subscriptions, support tickets, churn events, and feature usage records are linked to valid parent records.

6. Date columns were converted to proper datetime types.

7. The `feature_usage.usage_date` field has a lifecycle inconsistency: many usage records occur before the related account signup date or subscription start date. Because of this, `usage_date` will not be used for lifecycle sequencing or time-to-first-usage analysis.

8. Zero MRR/ARR values are explained by trial subscriptions. These rows should not be removed.

9. Numeric fields do not contain negative values.

10. Satisfaction scores are within the valid range from 1 to 5 when provided.

### Data Quality Decisions

- Keep missing `end_date` values because they represent active subscriptions.
- Keep missing `satisfaction_score` values because they represent customers who did not leave a rating.
- Keep zero MRR/ARR rows because they represent trial subscriptions.
- Keep all `feature_usage` rows, but do not use `usage_id` as the primary key.
- Do not use `feature_usage.usage_date` for lifecycle-based conclusions.
- Use `feature_usage` for product activity analysis: feature usage volume, error counts, and beta feature usage.

### Readiness for Next Step

The dataset is ready for cleaned CSV export and PostgreSQL setup, with documented limitations.

In [111]:
# Create cleaned copies of all tables

accounts_clean = accounts.copy()
subscriptions_clean = subscriptions.copy()
feature_usage_clean = feature_usage.copy()
support_tickets_clean = support_tickets.copy()
churn_events_clean = churn_events.copy()

# Add a technical primary key for feature_usage
# Reason: original usage_id is not fully unique
feature_usage_clean.insert(
    0,
    "feature_usage_id",
    range(1, len(feature_usage_clean) + 1)
)

print("Cleaned DataFrames created successfully.")

Cleaned DataFrames created successfully.


In [112]:
feature_usage_key_check = pd.DataFrame({
    "table_name": ["feature_usage_clean"],
    "total_rows": [len(feature_usage_clean)],
    "unique_feature_usage_id": [feature_usage_clean["feature_usage_id"].nunique()],
    "duplicate_feature_usage_id": [
        len(feature_usage_clean) - feature_usage_clean["feature_usage_id"].nunique()
    ],
    "missing_feature_usage_id": [feature_usage_clean["feature_usage_id"].isna().sum()]
})

feature_usage_key_check

,table_name,total_rows,unique_feature_usage_id,duplicate_feature_usage_id,missing_feature_usage_id
0,feature_usage_clean,25000,25000,0,0


In [113]:
import os

processed_path = "/content/processed"

os.makedirs(processed_path, exist_ok=True)

accounts_clean.to_csv(f"{processed_path}/accounts_clean.csv", index=False)
subscriptions_clean.to_csv(f"{processed_path}/subscriptions_clean.csv", index=False)
feature_usage_clean.to_csv(f"{processed_path}/feature_usage_clean.csv", index=False)
support_tickets_clean.to_csv(f"{processed_path}/support_tickets_clean.csv", index=False)
churn_events_clean.to_csv(f"{processed_path}/churn_events_clean.csv", index=False)

print("Cleaned CSV files exported successfully.")

Cleaned CSV files exported successfully.


In [114]:
processed_files = []

for file_name in os.listdir(processed_path):
    file_path = os.path.join(processed_path, file_name)

    temp_df = pd.read_csv(file_path)

    processed_files.append({
        "file_name": file_name,
        "row_count": len(temp_df),
        "column_count": temp_df.shape[1]
    })

processed_files = pd.DataFrame(processed_files).sort_values("file_name")

processed_files

,file_name,row_count,column_count
0,accounts_clean.csv,500,10
1,churn_events_clean.csv,600,9
3,feature_usage_clean.csv,25000,9
4,subscriptions_clean.csv,5000,14
2,support_tickets_clean.csv,2000,9


In [115]:
import shutil

shutil.make_archive(
    "/content/saas_processed_csv",
    "zip",
    processed_path
)

print("ZIP archive created: /content/saas_processed_csv.zip")

ZIP archive created: /content/saas_processed_csv.zip


In [116]:
from google.colab import files

files.download("/content/saas_processed_csv.zip")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>